src/Microsoft.ML.PipelineInference/AutoMlEngines/UniformRandomEngine.cs

```diff
@@ -30,8 +30,10 @@ public UniformRandomEngine(IHostEnvironment env)
             : base(env, env.Register("UniformRandomEngine(AutoML)"))
         {}
 
-        public override PipelinePattern[] GetNextCandidates(IEnumerable<PipelinePattern> history, int numberOfCandidates)
+        public override PipelinePattern[] GetNextCandidates(IEnumerable<PipelinePattern> history, int numberOfCandidates,
+            Dictionary<string, ColumnPurpose> colPurpose = null)
```

Comment:
>colPurpose [](start = 46, length = 10)

I am somewhat unenthusiastic about this, since it basically seems like role-mappings, except in reverse (which also makes it, incidentally, less capable). Is there any possibility of re-using that concept, rather than inventing another thing that acts more or less just like it?

Reply:
It seems the 'recipe framework'  added the concept of  'Purpose'.  Not quite sure why the decision was made at that time to  introduce Purpose as an additional concept,   instead of re-using the  concept of Roles and extending it.

<snip>
    public enum ColumnPurpose
    {
        Ignore = 0,
        Name = 1,
        Label = 2,
        NumericFeature = 3,
        CategoricalFeature = 4,
        TextFeature = 5,
        Weight = 6,
        GroupId = 7,
        ImagePath = 8
    }
</snip>

Is there a simple way to extend Roles to include purpose (e.g. 'TextFeature', 'Ignore')  ?  My understanding is that  it would be a bigger change  that warrants a separate design / PR.   

---
In reply to: [199661532](https://github.com/dotnet/machinelearning/pull/461#discussion_r199661532) [](ancestors = 199661532)

Reply:
You don't extend roles, you simply use them. If you want to introduce roles appropriate for your task, that is totally fine and supported. I wonder though if the issue is that you do not have an actual `ISchema` implementation we're working on?

Reply:
Is ColumnPurpose and Roles the same things? Feels to me they are not. There is a ColumnPurpose of "Date" but not such a Role, and Justin has ideas about adding more purposes. Purposes would be used to just communicate to Experts what transforms to suggest for the pipelines. Learners would not neccessarly need to know about them. 

---
In reply to: [200406010](https://github.com/dotnet/machinelearning/pull/461#discussion_r200406010) [](ancestors = 200406010)

Reply:
I don't think we have the actual ISchema implementation we're working on. AutoInference (aka Recipes) uses ColumnPurpose to get to there.  

---
In reply to: [200410413](https://github.com/dotnet/machinelearning/pull/461#discussion_r200410413) [](ancestors = 200410413,200406010)

Reply:
RecipeInference  uses  ColumnPurpose to get to the  Role-mappings.  

In the GUI users can modify the ColumnPurpose through the wizard.  We are adding arguments to the Macro so users can do the same when using the macro programmatically. 

@tfinley@gmail.com . Is the suggestion to completely do away with the notion of ColumnPurpose ?

---
In reply to: [200410413](https://github.com/dotnet/machinelearning/pull/461#discussion_r200410413) [](ancestors = 200410413,200406010)

Reply:
Made a couple of changes to address this:
- Dropped ‘Ignore’  from Arguments to the PipelineSweepMacro.
- Using RoleMappedData in the API   (not exposing ColumnPurpose anymore in the APIs)


---
In reply to: [200723741](https://github.com/dotnet/machinelearning/pull/461#discussion_r200723741) [](ancestors = 200723741,200410413,200406010)